coda-src/venus/9pfs.cc

```diff
@@ -705,10 +707,17 @@ int plan9server::recv_walk(unsigned char *buf, size_t len, uint16_t tag)
 
         DEBUG("9pfs: Twalk[%x] discarding after error wname[%u] = '%s'\n",
               tag, i, wname);
+        ::free(wname);
     }
 
-    /* report lookup errors only for the first path element */
-    if (i == 0 && conn->u.u_error) {
+    /* report lookup errors only for the first path element 
+       if (i == 0 && conn->u.u_error) { */
+    /* I don't understand the original intention here. 
+       In the case where nwname == 0, we would be checking 
+       u.u_error without having performed the lookup operation, 
+       so actually picking up the error from a prior operation. 
+       (AS) */
+    if (nwname > 0 && conn->u.u_error) {
```

Comment:
it tried to cover the case where nwname >= 1, if we fail the lookup on nwname[0] we return an error. Failures on nwname[1..n]  do not result in a returned error. See 6th paragraph in http://man.cat-v.org/inferno/5/walk

You are right that the nwname==0 case is not handled correctly, it should be similar to looking up '.', or possibly a dup(2) operation, we may have to explicitly set conn->u.u_error to 0 before looping through the names to make sure we don't trigger on a previous error.